crates/bdk/src/wallet/mod.rs

```diff
@@ -615,792 +457,977 @@ impl<D> Wallet<D> {
-            required_utxos,
-            optional_utxos,
-            fee_rate,
-            outgoing + fee_amount,
-            &drain_script,
-        )?;
-        fee_amount += coin_selection.fee_amount;
-        let excess = &coin_selection.excess;
+        // Same for the internal_policy path, if present
+        if params.change_policy != tx_builder::ChangeSpendPolicy::ChangeForbidden
+            && internal_policies
+                .iter()
+                .any(|(policy, _)| policy.requires_path())
+            && params.internal_policy_path.is_none()
+        {
+            return Err(Error::SpendingPolicyRequired(KeychainKind::Internal));
+        };
 
-        tx.input = coin_selection
-            .selected
+        let external_requirements = external_policies
             .iter()
-            .map(|u| bitcoin::TxIn {
-                previous_output: u.outpoint(),
-                script_sig: Script::default(),
-                sequence: n_sequence,
-                witness: Witness::new(),
+            .map(|(policy, _)| {
+                Ok(policy.get_condition(
+                    params
+                        .external_policy_path
+                        .as_ref()
+                        .unwrap_or(&BTreeMap::new()),
+                )?)
             })
-            .collect();
-
-        if tx.output.is_empty() {
-            // Uh oh, our transaction has no outputs.
-            // We allow this when:
-            // - We have a drain_to address and the utxos we must spend (this happens,
-            // for example, when we RBF)
-            // - We have a drain_to address and drain_wallet set
-            // Otherwise, we don't know who we should send the funds to, and how much
-            // we should send!
-            if params.drain_to.is_some() && (params.drain_wallet || !params.utxos.is_empty()) {
-                if let NoChange {
-                    dust_threshold,
-                    remaining_amount,
-                    change_fee,
-                } = excess
-                {
-                    return Err(Error::InsufficientFunds {
-                        needed: *dust_threshold,
-                        available: remaining_amount.saturating_sub(*change_fee),
-                    });
-                }
-            } else {
-                return Err(Error::NoRecipients);
-            }
-        }
+            .collect::<Result<Vec<_>, Error>>()?;
 
-        match excess {
-            NoChange {
-                remaining_amount, ..
-            } => fee_amount += remaining_amount,
-            Change { amount, fee } => {
-                if self.is_mine(&drain_script) {
-                    received += amount;
-                }
-                fee_amount += fee;
+        let internal_requirements = internal_policies
+            .iter()
+            .map(|(policy, _)| {
+                Ok(policy.get_condition(
+                    params
+                        .external_policy_path
```

Comment:
Shouldn't this logic be using `internal_policy_path`? I'm not very familiar with policy stuff but given that we now have multiple descriptors should we not keep track of multiple policy paths in the `TxParams` and not just the internal and external one?

Reply:
Good catch. This was a bug. And it doesn't get caught in tests because it doesn't test with complex multi-path change descriptors. We should add such tests.  Fixed in next push.